### PR TITLE
Support for optional/extra dependencies

### PR DESCRIPTION
**Pull request by alexjbinnie**
For things that install '.[tests]', allow packages to insert more options into the brackets

### CODE REVIEW

File: .github/workflows/mypy.yml [hunk 1]
```diff
@@ -124,7 +129,7 @@ jobs:
 
       - name: Install package
         working-directory: ./${{ inputs.path }}
-        run: python -m pip install ${{ inputs.pip_install_args }} .[tests]
+        run: python -m pip install ${{ inputs.pip_install_args }} .[${{ inputs.pip_optional_dependencies }}]
```
Comment (jandom) [inline .github/workflows/mypy.yml:132]: i would say this composes poorly – would rather have 

```
python -m pip install ${{ inputs.pip_install_args }} ${{ inputs.pip_optional_dependencies }}
```

with `pip_optional_dependencies` defaulting to "." – it's less magical in usage

Comment (alexjbinnie) [reply .github/workflows/mypy.yml]: Renamed to `package_specification` but did this, I agree it's nicer

**Review by daniellowtw — APPROVED** [review]
Looks backwards compatible to me.

**Comment by daniellowtw** [comment]
>allow packages to insert more options into the brackets

What are some examples here?
1. How does specifying multiple extras look like? Is this supported?

It looks like it's just doing pure string appending. This can probably cause clever string injection to add extra flags. I assume it no one would try to do that since this is an internal only workflow.

**Comment by alexjbinnie** [comment]
It's for `datagen` for now, basically if you want it to be installable with extra (i.e. `datagen` installs base packages, `datagen[all]` installs extras that allow more features but aren't required). When testing, you want all the features installed so you can test them all, so you need to install it as `.[all,tests]` instead of just `.[tests]`. 

Agree this is unsafe in the sense you can put extra args to pytest, but the same is true of the previous argument of install args.

**Pull request merged**